include/deal.II/base/thread_local_storage.h

```diff
@@ -19,22 +19,56 @@
 
 #  include <deal.II/base/config.h>
 
-DEAL_II_DISABLE_EXTRA_DIAGNOSTICS
-#  ifdef DEAL_II_WITH_THREADS
-#    include <tbb/enumerable_thread_specific.h>
-#  endif
-DEAL_II_ENABLE_EXTRA_DIAGNOSTICS
-
+#  include <deal.II/base/exceptions.h>
 
+#  include <list>
+#  include <map>
+#  include <shared_mutex>
+#  include <thread>
+#  include <vector>
 
 DEAL_II_NAMESPACE_OPEN
 
 /*!@addtogroup threads */
 /*@{*/
 
+#  ifndef DOXYGEN
+class LogStream;
+#  endif
 
 namespace Threads
 {
+#  ifndef DOXYGEN
+  namespace
+  {
+    /*
+     * Workaround: The standard unfortunately has an unfortunate design
+     * "flaw" in the std::is_copy_constructible type trait
+     * when it comes to STL containers and containing non-copyable objects
+     * T. The type trait is true even though any attempted invocation leads
+     * to a compilation error. Work around this issue by unpacking some
+     * commonly used containers:
+     */
+    template <typename T>
+    struct unpack_container
+    {
+      using type = T;
+    };
+
+    template <typename T, typename A>
+    struct unpack_container<std::vector<T, A>>
+    {
+      using type = T;
+    };
+
+    template <typename T, typename A>
+    struct unpack_container<std::list<T, A>>
+    {
+      using type = T;
+    };
+  } // namespace
```

Comment:
Could we put this in `Threads::internal`? anonymous namespaces in headers are problematic.

Reply:
Good point - I still have a reflexive aversion to using anonymous namespaces in headers but I suppose this is fine.

Reply:
@drwells I did put the `construct_element` overloads into an internal namespace because these will be potentially instantiated.

Reply:
If it makes @drwells uncomfortable, the namespace should probably be named. It also avoids mistakes at a later point if someone, for whatever reason, decides to put something in there without thinking this through.

Reply:
Fixed.